src/object.d

```diff
@@ -4956,3 +4956,315 @@ string _d_assert_fail(string comp, A, B)(A a, B b) @nogc @safe nothrow pure
+// Without this code will break as these hooks where previously exported
+// as having this safety level.
+T[] _d_arrayappendT(T)(ref T[] x, scope T[] y) @trusted
+{
+    alias func(T) = T[] function(ref T[] x, scope T[] y) @trusted pure nothrow;
+    auto f = cast(func!T)&_d_arrayappendT_!T;
+    return f(x, y);
+}
+
+T[] _d_arrayappendcTX(T)(ref T[] px, size_t n) @trusted
+{
+    alias func(T) = T[] function(ref T[] px, size_t n) @trusted pure nothrow;
+    auto f = cast(func!T)&_d_arrayappendcTX_!T;
+    return f(px, n);
+}
+
+T[] _d_arrayappendcT(T)(ref T[] px, scope T y) @trusted
+{
+    import core.internal.traits : Unqual;
+    import core.stdc.string : memcpy;
+    auto result = _d_arrayappendcTX(px, 1);
+    memcpy(cast(Unqual!T *)&result[$ - 1], cast(Unqual!T *)&y, T.sizeof);
+    return result;
+}
+
+/**
+ * Run postblit on `t` if it is a struct and needs it.
+ * Or if `t` is a array, run it on the children if they have a postblit.
+ */
+void __doPostblit(T)(auto ref T t)
+{
+    import core.internal.traits : hasElaborateCopyConstructor;
+
+    static if (is(T == struct)) {
+        // run the postblit function incase the struct has one
+        static if (__traits(hasMember, T, "__xpostblit") &&
+                // Bugzilla 14746: Check that it's the exact member of S.
+                __traits(isSame, T, __traits(parent, t.__xpostblit)))
+            t.__xpostblit();
+    }
+    else static if (is(T U : U[]) && hasElaborateCopyConstructor!U)
+    {
+        // only do a postblit if the `U` requires it.
+        foreach (ref el; t)
+            __doPostblit(el);
+    }
+}
+
+T[] _d_arrayappendTTrace(T)(string file, int line, string funcname, ref T[] x, scope T[] y) @trusted pure
+{
+    import rt.profilegc : accumulate;
+
+    string typename = T.stringof;
+    version (tracegc)
+    {
+        printf("_d_arrayappendT type = %.*s x = %llu y = %llu file = '%.*s' line = %d function = '%.*s'\n",
+            cast(int)typename.length, typename.ptr,
+            cast(ulong)x.length,
+            cast(ulong)y.length,
+            file.length, file.ptr,
+            line,
+            funcname.length, funcname.ptr
+            );
+    }
+    accumulate(file, line, funcname, typename, T.sizeof * y.length);
+    return _d_arrayappendT(x, y);
+}
+
+/**
+ * Append y[] to array x[]
+ */
+T[] _d_arrayappendT_(T)(ref T[] x, scope T[] y) @trusted
+{
+    import core.stdc.string;
+    import core.internal.traits : Unqual;
+    alias unqualT = Unqual!T;
+
+    auto length = x.length;
+    auto sizeelem = T.sizeof;              // array element size
+    _d_arrayappendcTX(x, y.length);
+    memcpy(cast(Unqual!T *)&x[length], cast(Unqual!T *)&y[0], y.length * sizeelem);
+
+    // do postblit
+    __doPostblit(cast(Unqual!T[])x[length .. length + y.length]);
+    return x;
+}
+
+T[] _d_arrayappendcTXTrace(T)(string file, int line, string funcname, ref T[] px, size_t n) @trusted pure
+{
+    import rt.profilegc : accumulate;
```

Comment:
Note that the implementation of the tracing functions has changed recently, see https://github.com/dlang/druntime/pull/2607

I wouldn't mind if the trace-functions are left unmodified, it's a half-baked feature anyway and slows down execution in accumulate.

Reply:
The problem is that the new entrypoints for the hooks are templates, so I need to manually write a trace entrypoint. It also need to be in object.d so the compiler can reference it from the lowering code. https://github.com/dlang/dmd/pull/9982/files#diff-0c1aa1ff99402eb87e11cc3fa1b80da0